test/test_framework.py

```diff
@@ -750,6 +750,25 @@ def test_ban_concurrent_frameworks(self):
         self.assertIn('database is locked', str(cm.exception))
         f.close()
 
+    def test_snapshot_saving_restricted_to_simple_types(self):
+        # this can not be saved, as it has not simple types!
+        to_be_saved = {"foo": 1, "bar": TestFramework}
+
+        class FooEvent(EventsBase):
+            def snapshot(self):
+                return to_be_saved
+
+        handle = Handle(None, "a_foo", "some_key")
+        event = FooEvent()
+
+        framework = self.create_framework()
+        framework.register_type(FooEvent, None, handle.kind)
+        with self.assertRaises(ValueError) as cm:
+            framework.save_snapshot(event)
+        expected = "Can not save the following data, must contain only simple types: {!r}".format(
+            to_be_saved)
```

Comment:
Given that the point of this is to make errors to users understandable, I feel like spelling this out rather than saying 'it matches the format string' makes it clearer to developers looking at the test to say "yes, this is an error I'd like to read as a developer'.
In particular, I feel that #194 is about not knowing what object was being marshalled, not just that it wasn't simple types.
Eg, he was hoping to get the FooEvent, and possibly the actual value that couldn't be marshalled.

Reply:
There I spelled the whole message.

Good idea to include which object was being handled when the marshaling problem appeared. Did that.

Showing *only* the problematic value is too complicated: we would need to walk on the whole structure, checking everything, etc. Showing the whole structure should be enough.

Reply:
(Yes, both the object being marshalled and the repr(value) that couldn't be marshalled.

I don't know the reason why only simple types are allowed here, so maybe a comment while you are at it :) Storing pickles rather than yaml or json seems weird when you are restricting things to simple data structures. The advantage of pickle is you don't need snapshot and restore methods at all, serializing anything that is pickleable per the Python standard protocol.

Reply:
We allow only simple types because we're actually pickling the data to store it. And then, at load time, we unpickle it. 

If we would allow custom objects, we're giving the developer the oppurtinity to shoot their feet, as if they change the name of the class of the custom object they used, all further events loading will fail! Unpickling needs access to the original class!

That said, this is not we will NEVER allow. But it's easier to relax this limitation in the future if really needed, than restrict later something originally too permissive.

BTW, we're using pickle and not marshal, because marshal format changes in time, so pickle guarantees us that the event stored with a version of Python will be able to be read in the future.